### PR TITLE
Feature/OnRCStatus `allowed` parameter 

### PR DESCRIPTION
**Pull request by KhrystynaDubovyk**
This implementation is based on OnRCStatus notification, so this PR should be merged after #75 

Technical task : [#2244](https://github.com/smartdevicelink/sdl_core/issues/2244) 

This pull request contains the implementation of Remote Control - Update OnRCStatus with a new allowed parameter
proposal:[0172-onRcStatus-allowed.md](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0172-onRcStatus-allowed.md)

Test scripts: smartdevicelink/sdl_atf_test_scripts#1952
For testing use the following test set:
https://github.com/smartdevicelink/sdl_atf_test_scripts/blob/feature/onRcStatus_allowed/test_sets/rc_OnRCStatus.txt

### CODE REVIEW

File: detailed_docs/RC/OnRCStatus.md [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+b.	RC functionality is allowed on HMI  
+
+c.	All modules are free  
+
+d.	RC functionality is allowed by policies  
+
+_Steps:_
+
+1.	RC application starts registration (upon a OnButtonPress or SetInteriorVehicleData request)
```
Comment (yang1070) [inline detailed_docs/RC/OnRCStatus.md]: why mention `upon a OnButtonPress or SetInteriorVehicleData request`?

File: detailed_docs/RC/OnRCStatus.md [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+_Expected:_   
+1. SDL sends OnRCStatus with all modules in `freeModules` parameter to HMI
```
Comment (yang1070) [inline detailed_docs/RC/OnRCStatus.md]: + empty allocatedModules

File: detailed_docs/RC/OnRCStatus.md [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+b.	RC functionality is allowed on HMI  
+
+c.	All modules are free  
+
+d.	RC functionality is allowed by policies  
+
+e.	RC app_1 is registered 
+
+_Steps:_
+1.	RC app_1 sends control request to module_1
```
Comment (yang1070) [inline detailed_docs/RC/OnRCStatus.md]: `RC app_1 sends control request to module_1`  (upon a OnButtonPress or SetInteriorVehicleData request)?

File: detailed_docs/RC/resource_allocation.md [hunk 4]
```diff
@@ -14,33 +14,40 @@ _Steps:_
 
 1. User chose to allow RC functionality via HMI and specified access mode as AUTO_ALLOW or didn't specify the access mode at all
 2. HMI sends notification to SDL that Remote control functionality is allowed with or without specified access mode
-3. RC_app_1 is in HMILevel FULL and is in control of the module_1 (sent request to either change settings or emulate button press)
-4. User activates RC_app_2 (app_2 is in FULL HMI level now)
-5. RC_app_2 sends control request to module_1 (sent request to either change settings or emulate button press)
+3. RC_app_1 is in HMILevel FULL sends request to allocate the module _1 (either  to change settings or emulate button press)
+4. User activates RC_app_2 (app_2 is in FULL HMI level now)	+4. RC_app_1 is in control of the the module_1
```
Comment (yang1070) [inline detailed_docs/RC/resource_allocation.md]: two 4?

File: detailed_docs/RC/resource_allocation.md [hunk 5]
```diff
@@ -14,33 +14,40 @@ _Steps:_
...
+5. RC_app_2 sends control request to module_1 (sent request to either change settings or emulate button press)	+5. SDL sends OnRCStatus to HMI and to registered application app1 and app2 with module_1 in section `allocatedModules` and remained modules in `freeModules`
```
Comment (yang1070) [inline detailed_docs/RC/resource_allocation.md]:
```
SDL sends OnRCStatus to HMI and to registered application app1 and app2 with module_1 in section `allocatedModules` and remained modules in `freeModules` 
```
`module_1` shall be in only one `allocatedModules`  in notification to the app.
in this case, the notification to app1 include `module_1`  in `allocatedModules` , the notification to app2 does not include `module_1`  in `allocatedModules`

Comment (yang1070) [inline detailed_docs/RC/resource_allocation.md]: two 5

File: detailed_docs/RC/resource_allocation.md [hunk 6]
```diff
@@ -14,33 +14,40 @@ _Steps:_
 
 1. User chose to allow RC functionality via HMI and specified access mode as AUTO_ALLOW or didn't specify the access mode at all
 2. HMI sends notification to SDL that Remote control functionality is allowed with or without specified access mode
-3. RC_app_1 is in HMILevel FULL and is in control of the module_1 (sent request to either change settings or emulate button press)
-4. User activates RC_app_2 (app_2 is in FULL HMI level now)
-5. RC_app_2 sends control request to module_1 (sent request to either change settings or emulate button press)
+3. RC_app_1 is in HMILevel FULL sends request to allocate the module _1 (either  to change settings or emulate button press)
+4. User activates RC_app_2 (app_2 is in FULL HMI level now)	+4. RC_app_1 is in control of the the module_1
+5. RC_app_2 sends control request to module_1 (sent request to either change settings or emulate button press)	+5. SDL sends OnRCStatus to HMI and to registered application app1 and app2 with module_1 in section `allocatedModules` and remained modules in `freeModules`  
+6. User activates RC_app_2 (app_2 is in FULL HMI level now)
```
Comment (yang1070) [inline detailed_docs/RC/resource_allocation.md]: same as 4?

File: detailed_docs/RC/resource_allocation.md [hunk 7]
```diff
@@ -14,33 +14,40 @@ _Steps:_
...
+4. User activates RC_app_2 (app_2 is in FULL HMI level now)	+4. RC_app_1 is in control of the the module_1
+5. RC_app_2 sends control request to module_1 (sent request to either change settings or emulate button press)	+5. SDL sends OnRCStatus to HMI and to registered application app1 and app2 with module_1 in section `allocatedModules` and remained modules in `freeModules`  
+6. User activates RC_app_2 (app_2 is in FULL HMI level now)  
+7. RC_app_2 sends control request to module_1 (sent request to either change settings or emulate button press)
 
 _Expected:_
 
-6. RC_app_2 gets in control of the module_1
+8. RC_app_2 gets in control of the module_1  
+9. SDL sends OnRCStatus to HMI and to registered application app1 and app2 with module_1 in section allocatedModules
```
Comment (yang1070) [inline detailed_docs/RC/resource_allocation.md]: in this case, the notification to app1 does not include module_1 in allocatedModules , the notification to app2  include module_1 in allocatedModules

File: detailed_docs/RC/resource_allocation.md [hunk 8]
```diff
@@ -150,32 +187,37 @@ _Steps:_
 
 1. User disables RC functionality
 2. HMI sent notification to SDL that RC functionality is disabled
-3. SDL assigns HMILevel NONE to RC_app_1 and RC_app_2 and unsubscribes applications from module_1
+3. SDL assigns HMILevel NONE to RC_app_1 and RC_app_2 and unsubscribes applications from module_1
```
Comment (yang1070) [inline detailed_docs/RC/resource_allocation.md]: please see https://github.com/smartdevicelink/sdl_core/issues/2173 and https://github.com/smartdevicelink/sdl_evolution/issues/531, we require keep the current HMI level

File: detailed_docs/RC/resource_allocation.md [hunk 9]
```diff
@@ -150,32 +187,37 @@ _Steps:_
...
 5. User activates RC_app_2, module_1 gets in "free" state again
 6. RC_app_2 requests the control of module_1
 
 _Expected:_
 
-7. RC_app_2 gets in control of module_1 without asking a driver
-8. User activates RC_app_1 and RC_app_1 requests the control of module_1
-9. SDL sends the request to HMI to display permission prompt to the driver
+7. RC_app_2 gets in control of module_1 without asking a driver  
+8. SDL sends OnRCStatus to RC_app_1 and RC_app_2 and to HMI with module_1 in `allocatedModules` parameter.
```
Comment (yang1070) [inline detailed_docs/RC/resource_allocation.md]: module_1 can only be in one allocatedModules

File: detailed_docs/RC/resource_allocation.md [hunk 10]
```diff
@@ -211,35 +254,37 @@ _Expected:_
 
 _Expected:_
 
-1.a.3 SDL assigns HMILevel NONE to RC_app_1, releases module_1 from RC_app_1 control and unsubscribes RC_app_1 from receiving notifications about module_1 settings changed
+1.a.3 SDL assigns HMILevel NONE to RC_app_1, releases module_1 from RC_app_1 control and unsubscribes RC_app_1 from receiving notifications about module_1 settings changed  
 
-1.a.4 RC_app_2 gets in control of module_1
+1.a.4 SDL sends OnRCStatus to HMI, RC_app_1 and RC_app_2 with module_1 in `freeModules` parameter
```
Comment (yang1070) [inline detailed_docs/RC/resource_allocation.md]: app_1 is in NONE. SDL shall not send OnRCStatus to app_1

File: detailed_docs/RC/resource_allocation.md [hunk 11]
```diff
@@ -259,14 +304,18 @@ _Expected:_
 
 1.c.3 SDL assigns HMILevel NONE to all applications registered from deviceID signed in deviceInfo and unsubscribes all applications of this device from all modules
 
-1.c.4 module_1 is not allocated to any application
+1.c.4 module_1 is not allocated to any application  
+
+1.c.5  SDL sends OnRCStatus to HMI, RC_app_1 and RC_app_2 with module_1 in `freeModules` parameter
```
Comment (yang1070) [inline detailed_docs/RC/resource_allocation.md]: SDL  send OnRCStatus to apps in NONE?

File: detailed_docs/RC/resource_allocation.md [hunk 12]
```diff
@@ -259,14 +304,18 @@ _Expected:_
...
 _Exception 3.1_
 
 1.c.1.1 SDL received OnAllowSDLFunctionality (allowed: FALSE) without specified deviceInfo
 
 1.c.1.2 SDL assigns HMILevel NONE to all applications registered from all devices and unsubscribes all applications from all modules
 
-1.c.1.3 module_1 is not allocated to any application
+1.c.1.3 module_1 is not allocated to any application  
+
+1.c.1.4 SDL sends OnRCStatus to HMI, RC_app_1 and RC_app_2 with module_1 in `freeModules` parameter
```
Comment (yang1070) [inline detailed_docs/RC/resource_allocation.md]: SDL  send OnRCStatus to apps in NONE?

File: detailed_docs/TRS/RC/OnRCStatus.md [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+SDL must  
+send OnRCStatus notification with RC modules in 'freeModules' and 'allocatedModules' parameters to HMI  
+send OnRCStatus notification with 'allowed = true', 'freeModules' and 'allocatedModules' parameters to mobile application 
+
+2.
+In case   
+- RC functionality is disallowed on HMI  
+- all RC modules are free
+
+AND RC application sends ButtonPress or SetInteriorVehicleData request
```
Comment (yang1070) [inline detailed_docs/TRS/RC/OnRCStatus.md]: the request to ButtonPress or SetInteriorVehicleData shall be reject, no need to send  OnRCStatus for the requests

File: detailed_docs/TRS/RC/OnRCStatus.md [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+send OnRCStatus notification with 'allowed = true', 'freeModules' and 'allocatedModules' parameters to mobile application 
+
+2.
+In case   
+- RC functionality is disallowed on HMI  
+- all RC modules are free
+
+AND RC application sends ButtonPress or SetInteriorVehicleData request 
+
+SDL must
```
Comment (yang1070) [inline detailed_docs/TRS/RC/OnRCStatus.md]: no need to send OnRC Status here

File: detailed_docs/TRS/RC/OnRCStatus.md [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+6.  
+In case  
+- RC functionality is allowed on HMI   
+- all RC modules are free
+- RC app_1 is registered with SDL
+
+AND RC app_2 starts registration
+
+SDL must  
+send OnRCStatus notification with all modules in `freeModules` parameter to HMI and to RC app_1, RC app_2 with all modules in `freeModules` parameter.
```
Comment (yang1070) [inline detailed_docs/TRS/RC/OnRCStatus.md]: why send to app_1? there is nothing change
 shall send only to the newly registered app_2, I think HMI also not need it

File: detailed_docs/TRS/RC/OnRCStatus.md [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+SDL must  
+NOT send OnRCStatus to HMI and to registered non-RC application
+
+4. 
+In case  
+- RC functionality is allowed on HMI  
+- all RC modules are free
+- RC application is registered with SDL
+
+AND RC application allocates module_1
```
Comment (yang1070) [inline detailed_docs/TRS/RC/OnRCStatus.md]: app_1?

File: detailed_docs/TRS/RC/OnRCStatus.md [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+- all RC modules are free
+- RC app_1 and RC app_2 are registered with SDL
+
+AND RC app_1 allocates module_1  
+
+SDL must  
+send OnRCStatus to HMI and to RC app_1, RC app_2 with module_1 in section `allocatedModules`  
+and remained modules in `freeModules`
+
+ 8.
```
Comment (yang1070) [inline detailed_docs/TRS/RC/OnRCStatus.md]: similar to case4, instead of app_1 control module_1, now app_2 control module_1

**Review by yang1070 — CHANGES_REQUESTED** [review]
Please update

File: detailed_docs/RC/resource_allocation.md (hunk 4 above)
Comment (KhrystynaDubovyk) [reply detailed_docs/RC/resource_allocation.md]: @yang1070, thank you. 
please see updates https://github.com/smartdevicelink/sdl_requirements/pull/89/commits/bcefffcf4992deb88e3ab84717ac519cbd318197

File: detailed_docs/RC/resource_allocation.md (hunk 5 above)
Comment (KhrystynaDubovyk) [reply detailed_docs/RC/resource_allocation.md]: @yang1070, thank you. 
please see updates https://github.com/smartdevicelink/sdl_requirements/pull/89/commits/bcefffcf4992deb88e3ab84717ac519cbd318197

Comment (KhrystynaDubovyk) [reply detailed_docs/RC/resource_allocation.md]: @yang1070, thank you. 
please see updates https://github.com/smartdevicelink/sdl_requirements/pull/89/commits/bcefffcf4992deb88e3ab84717ac519cbd318197

File: detailed_docs/RC/resource_allocation.md (hunk 6 above)
Comment (KhrystynaDubovyk) [reply detailed_docs/RC/resource_allocation.md]: @yang1070, thank you. 
please see updates https://github.com/smartdevicelink/sdl_requirements/pull/89/commits/bcefffcf4992deb88e3ab84717ac519cbd318197

File: detailed_docs/RC/resource_allocation.md (hunk 7 above)
Comment (KhrystynaDubovyk) [reply detailed_docs/RC/resource_allocation.md]: @yang1070, thank you. 
please see updates https://github.com/smartdevicelink/sdl_requirements/pull/89/commits/bcefffcf4992deb88e3ab84717ac519cbd318197

File: detailed_docs/RC/resource_allocation.md (hunk 9 above)
Comment (KhrystynaDubovyk) [reply detailed_docs/RC/resource_allocation.md]: @yang1070, thank you. 
please see updates https://github.com/smartdevicelink/sdl_requirements/pull/89/commits/bcefffcf4992deb88e3ab84717ac519cbd318197

File: detailed_docs/RC/resource_allocation.md (hunk 8 above)
Comment (KhrystynaDubovyk) [reply detailed_docs/RC/resource_allocation.md]: @yang1070, thank you. these changes will be added after accepting the proposal in the corresponding feature branch with a separate PR.

File: detailed_docs/RC/OnRCStatus.md (hunk 1 above)
Comment (KhrystynaDubovyk) [reply detailed_docs/RC/OnRCStatus.md]: @yang1070, please see the updates in https://github.com/smartdevicelink/sdl_requirements/pull/89/commits/2e9692b47878e6f1b766e7fbd380673c29af74db

File: detailed_docs/RC/OnRCStatus.md (hunk 2 above)
Comment (KhrystynaDubovyk) [reply detailed_docs/RC/OnRCStatus.md]: @yang1070, please see the updates in https://github.com/smartdevicelink/sdl_requirements/pull/89/commits/2e9692b47878e6f1b766e7fbd380673c29af74db

File: detailed_docs/RC/OnRCStatus.md (hunk 3 above)
Comment (KhrystynaDubovyk) [reply detailed_docs/RC/OnRCStatus.md]: @yang1070, please see the updates in https://github.com/smartdevicelink/sdl_requirements/pull/89/commits/2e9692b47878e6f1b766e7fbd380673c29af74db

File: detailed_docs/TRS/RC/OnRCStatus.md (hunk 13 above)
Comment (KhrystynaDubovyk) [reply detailed_docs/TRS/RC/OnRCStatus.md]: @yang1070, thank you
please, see updates https://github.com/smartdevicelink/sdl_requirements/pull/89/commits/9a23cb27fb2e940fd3601f71687d8665ce8618e6

File: detailed_docs/TRS/RC/OnRCStatus.md (hunk 14 above)
Comment (KhrystynaDubovyk) [reply detailed_docs/TRS/RC/OnRCStatus.md]: @yang1070, thank you
please, see updates https://github.com/smartdevicelink/sdl_requirements/pull/89/commits/9a23cb27fb2e940fd3601f71687d8665ce8618e6

File: detailed_docs/TRS/RC/OnRCStatus.md (hunk 15 above)
Comment (KhrystynaDubovyk) [reply detailed_docs/TRS/RC/OnRCStatus.md]: @yang1070, thank you
please, see updates https://github.com/smartdevicelink/sdl_requirements/pull/89/commits/9a23cb27fb2e940fd3601f71687d8665ce8618e6

File: detailed_docs/TRS/RC/OnRCStatus.md [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+- all RC modules are free
+- RC app_1 is registered with SDL
+
+AND RC app_2 starts registration
+
+SDL must  
+send OnRCStatus notification with all modules in `freeModules` parameter to HMI and to RC app_1, RC app_2 with all modules in `freeModules` parameter.  
+send and OnRCStatus notification with 'allowed = true' and all modules in `freeModules` parameter to RC app_1, RC app_2 
+
+7.
```
[the remark this answers is not in the document]
Comment (KhrystynaDubovyk) [reply detailed_docs/TRS/RC/OnRCStatus.md]: @yang1070, thank you
I have rewritten case#4 is with 1 app,  
case#7 with 2 apps https://github.com/smartdevicelink/sdl_requirements/pull/89/commits/9a23cb27fb2e940fd3601f71687d8665ce8618e6

File: detailed_docs/TRS/RC/OnRCStatus.md (hunk 16 above)
Comment (KhrystynaDubovyk) [reply detailed_docs/TRS/RC/OnRCStatus.md]: @yang1070 yes, app_1. updated https://github.com/smartdevicelink/sdl_requirements/commit/9a23cb27fb2e940fd3601f71687d8665ce8618e6

**Review by LitvinenkoIra — APPROVED** [review]

File: detailed_docs/RC/OnRCStatus.md [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+
+e.	RC app_1 is registered
+
+_Steps:_
+1.	RC app_1 sends control request to module_1 (upon a OnButtonPress or SetInteriorVehicleData request)
+
+2. RC app_1 allocates module_1
+
+_Expected:_
+3. SDL sends OnRCStatus notification to HMI with module_1 in `allocatedModules` parameter and remained modules in `freeModules`
```
Comment (JackLivio) [inline detailed_docs/RC/OnRCStatus.md]: `remaining`

File: detailed_docs/RC/OnRCStatus.md [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+4. SDL sends OnRCStatus notification to RC app_1 with (allowed = true), module_1 in `allocatedModules` parameter and remained modules in `freeModules`
```
Comment (JackLivio) [inline detailed_docs/RC/OnRCStatus.md]: remaining**

File: detailed_docs/RC/OnRCStatus.md [hunk 21]
```diff
@@ -0,0 +1,147 @@
...
+
+**Alternative flow 1**
+1.a.1 RC app_1 already allocates module_1
+1.a.2 RC app_1 sends control request to module_1 one more time
+1.a.3 SDL does not send OnRCStatus
+
+**Alternative flow 2**
```
Comment (JackLivio) [inline detailed_docs/RC/OnRCStatus.md]: I am a little unclear on what this flow is describing.

File: detailed_docs/RC/OnRCStatus.md (hunk 19 above)
Comment (LitvinenkoIra) [reply detailed_docs/RC/OnRCStatus.md]: @JackLivio fixed 
69eb722

File: detailed_docs/RC/OnRCStatus.md (hunk 20 above)
Comment (LitvinenkoIra) [reply detailed_docs/RC/OnRCStatus.md]: @JackLivio fixed 
69eb722

File: detailed_docs/RC/OnRCStatus.md (hunk 21 above)
Comment (LitvinenkoIra) [reply detailed_docs/RC/OnRCStatus.md]: @JackLivio when we try to acquire some module (upon a OnButtonPress or SetInteriorVehicleData request) and HMI doesn't response or response with invalid values of parameters -> module doen't not acquired -> SDL does not sends OnRCStatus

**Pull request merged**